elements/pl-truth-table/pl-truth-table.py

```diff
@@ -0,0 +1,567 @@
+
+                        if (i // power) % 2 != 0:
+                            val = 1
+
+                        values.append(val)
+                        var_vals[display_variables[j]["html"]] = val
+
+                    keyed_row["values"] = values
+                    display_rows.append(keyed_row)
+                    for j in range(len(answer_columns)):
+                        correct_answers.append(str(bool(eval(expressions[j]))))
+    elif len(custom_rows) == 0:
+        num_rows = default_num_rows
+
+        for i in range(num_rows):
+            keyed_row = {
+                "key": str(i),
+            }
+            values = []
+
+            for j in range(num_vars):
+                power = pow(2, num_vars - j - 1)
+                val = 0
+
+                if (i // power) % 2 != 0:
+                    val = 1
+
+                values.append(val)
+                var_vals[display_variables[j]["html"]] = val
+
+            keyed_row["values"] = values
+            display_rows.append(keyed_row)
+            for j in range(len(answer_columns)):
+                correct_answers.append(str(bool(eval(expressions[j]))))
+    else:
+        num_rows = len(custom_rows)
+
+        for i in range(num_rows):
+            keyed_row = {"key": str(i)}
+
+            row_info = custom_rows[i].split()
+            values = []
+            var = ""
+            val = 0
+
+            for j in range(len(row_info)):
+                values.append(int(row_info[j]))
+                var_vals[display_variables[j]["html"]] = int(row_info[j])
+
+            keyed_row["values"] = values
+            display_rows.append(keyed_row)
+            for j in range(len(answer_columns)):
+                correct_answers.append(str(bool(eval(expressions[j]))))
+
+    data["params"][name] = (display_variables, display_rows, display_ans_columns)
+    data["correct_answers"][name] = correct_answers
+
+
+def parse(element_html, data):
+    element = lxml.html.fragment_fromstring(element_html)
+    name = pl.get_string_attrib(element, "answers-name")
+    display_variables, display_rows, display_ans_columns = data["params"].get(name)
+    submitted_answers = data["submitted_answers"].get(name, [])
+
+    for i in range(len(display_rows)):
+        expected_html_name = get_form_name(name, i)
+        try:
+            student_answer = submitted_answers[i]
+        except (ValueError, TypeError):
+            data["format_errors"][
+                expected_html_name
+            ] = "The submitted answer is not a legal option."
+            continue
+
+        # A blank is a valid submission from the HTML, but causes a format error.
+        if student_answer == -1:
+            data["format_errors"][
+                expected_html_name
+            ] = "The submitted answer was left blank."
+        elif student_answer is None:
+            data["format_errors"][expected_html_name] = "No answer was submitted."
+
+
+def render(element_html, data):
```

Comment:
There is a huge amount of code duplication between the `question`, `answer` and `submission` panels. I think a lot of this code can be factored out into a function for code reuse.